ext/reflection/php_reflection.c

```diff
@@ -1372,33 +1330,81 @@ static parameter_reference *_reflection_param_get_default_param(INTERNAL_FUNCTIO
 	}
 
 	param = intern->ptr;
-	if (param->fptr->type != ZEND_USER_FUNCTION) {
-		zend_throw_exception_ex(reflection_exception_ptr, 0, "Cannot determine default value for internal functions");
-		return NULL;
-	}
 
 	return param;
 }
 /* }}} */
 
-/* {{{ _reflection_param_get_default_precv */
-static zend_op *_reflection_param_get_default_precv(INTERNAL_FUNCTION_PARAMETERS, parameter_reference *param)
+/* {{{ _reflection_param_get_default_arg_info */
+static int _reflection_param_get_default_arg_info(zend_arg_info *arg_info, zval *default_value_zval)
 {
-	zend_op *precv;
+	if (arg_info->default_value == NULL) {
+		zend_throw_exception_ex(reflection_exception_ptr, 0, "Internal error: Failed to retrieve the default value");
+		return FAILURE;
+	}
+
+	zval code_zv;
+	zend_bool original_in_compilation;
+	uint32_t original_compiler_options;
+	zend_lex_state original_lex_state;
+	zend_ast *ast;
+	zend_arena *ast_arena;
+	int success = FAILURE;
+
+	smart_str code = {0};
+	smart_str_appends(&code, "<?php ");
+	smart_str_appends(&code, arg_info->default_value);
+	smart_str_appendc(&code, ';');
+	smart_str_0(&code);
+
+	ZVAL_STR_COPY(&code_zv, code.s);
+	smart_str_free(&code);
+
+	original_in_compilation = CG(in_compilation);
+	original_compiler_options = CG(compiler_options);
+	zend_save_lexical_state(&original_lex_state);
+
+	CG(in_compilation) = 1;
+	CG(compiler_options) |= ZEND_COMPILE_NO_CONSTANT_SUBSTITUTION | ZEND_COMPILE_NO_PERSISTENT_CONSTANT_SUBSTITUTION;
+
+	if (zend_prepare_string_for_scanning(&code_zv, "") == SUCCESS) {
+		CG(ast) = NULL;
+		CG(ast_arena) = zend_arena_create(1024 * 32);
+		LANG_SCNG(yy_state) = yycINITIAL;
+
+		if (zendparse() != 0) {
+			zend_ast_destroy(CG(ast));
+			zend_arena_destroy(CG(ast_arena));
+			CG(ast) = NULL;
+		}
+	}
 
-	if (param == NULL) {
-		return NULL;
+	ast = CG(ast);
+	ast_arena = CG(ast_arena);
+
+	if (ast) {
+		zend_ast_list *statement_list = zend_ast_get_list(ast);
+		zend_ast *const_expression_ast = statement_list->child[0];
+		if (const_expression_ast) {
+			zend_const_expr_to_zval(default_value_zval, const_expression_ast);
+			success = SUCCESS;
+		}
 	}
 
-	precv = _get_recv_op((zend_op_array*)param->fptr, param->offset);
-	if (!precv || precv->opcode != ZEND_RECV_INIT || precv->op2_type == IS_UNUSED) {
+	zend_restore_lexical_state(&original_lex_state);
+	CG(compiler_options) = original_compiler_options;
+	CG(in_compilation) = original_in_compilation;
+
+	zval_dtor(&code_zv);
+	zend_ast_destroy(ast);
+	zend_arena_destroy(ast_arena);
```

Comment:
This seems like a very complex way to go about it. Wouldn't just `zend_eval_string*`ing it work?

Reply:
Embarrassingly complex, I know... :/

I've just tried `zend_eval_string(ZSTR_VAL(arg_info->default_value), default_value_zval, "");`, but no success yet because of problems with constants like `Undefined class constant ...`.

Reply:
Hm... this is tricky.